NMS-14182: Clean up Alarmd doc formatting (#4557)
OpenNMS/opennms: docs/modules/operation/pages/alarms/configuring-alarms.adoc

*NOTE*
See also: <<events/anatomy-events.adoc#ga-events-anatomy-of-an-event, anatomy of an event>>

== The reduction-key

Alarmd is designed to reduce multiple occurrences of an alarm into a single alarm.

> The critical attribute when defining the alarm-data of an event, is the reduction-key.

Reviewer: ```suggestion
The critical attribute when defining the alarm-data of an event is the reduction-key.
```

This attribute can contain literal strings as well as references to properties (fields and parameters) of the event.

> The purpose of the reduction-key is to uniquely identify the signature of a problem and, as such, is used to reduce (de-duplicate) events so that only one problem is instantiated.

Reviewer: ```suggestion
The reduction-key uniquely identifies the signature of a problem and, as such, is used to reduce (de-duplicate) events so that only one problem is instantiated.
```

> Most commonly, the event's identifier (UEI) is used as the left most (least significant) portion of the reduction-key, followed by other properties of the event from least to most significant and, traditionally, separated with the literal `:`.

Reviewer: ```suggestion
Most commonly, the event's identifier (UEI) is used as the left-most (least significant) portion of the reduction-key, followed by other properties of the event from least to most significant and, traditionally, separated with the literal `:`.
```

----
With this reduction-key, a single alarm would be instantiated for all nodes that were determined by the Poller to be down.
There would be a single alarm with the count representing the number of nodes down.

> However, the UEI `uei.opennms.org/nodes/nodeUp` would *not* be a good 'pair wise' reduction-key for resolving this alarm as it would take only a single 'node up' to clear all nodes down tracked with this single alarm configuration.

Author: I don't think "pairwise" (with or without spaces) appears in product at all.  I think there is just one other reference outside the docs, and that's in a comment in the default opennms.properties file.

```suggestion
However, the UEI `uei.opennms.org/nodes/nodeUp` would *not* be a good 'pairwise' reduction-key for resolving this alarm as it would take only a single 'node up' to clear all nodes down tracked with this single alarm configuration.
```

====

> == The alarm-type attribute

Reviewer: ```suggestion
== Alarm-type attribute
```

The second most critical attribute is the alarm-type.
There are currently three types of alarms:

* problem (1)
* resolution (2)
* notification (3)

> The alarm-type attribute helps Alarmd with pair-wise resolution, matching the resolution events to problem events.

Reviewer: ```suggestion
The alarm-type attribute helps Alarmd with pairwise resolution, matching the resolution events to problem events.
```

> NOTE: If an alarm transitions from an alarm-type 2 back to alarm-type 1 the Severity will be set to the most recent event's value.

Reviewer: ```suggestion
NOTE: If an alarm transitions from an alarm-type 2 back to alarm-type 1, the severity will be set to the most recent event's value.
```

> == The clear-key attribute

Reviewer: ```suggestion
== Clear-key attribute
```

> This attribute is used in the pair-wise correlation feature of Alarmd.

Reviewer: ```suggestion
This attribute is used in the pairwise correlation feature of Alarmd.
```

> When configuring a resolution alarm, set this attribute to match the reduction-key of a the corresponding problem alarm.

Reviewer: ```suggestion
When configuring a resolution alarm, set this attribute to match the reduction-key of the corresponding problem alarm.
```

> == The auto-clean attribute

Reviewer: ```suggestion
== Auto-clean attribute
```

> This attribute instructs Alarmd to only retain the most recent event reduced into an alarm.

Reviewer: ```suggestion
This attribute instructs Alarmd to retain only the most recent event reduced into an alarm.
```

For alarms that are super chatty, this is a way to reduce the size of the most recent events in the database.

WARNING: Do not use this feature with alarms that have pair-wise correlation (matching problems with resolutions).

> == The update-field element

Reviewer: ```suggestion
== Update-field element
```

* acktime
* ackuser

== Pairwise Correlation
Alarmd is intrinsically designed to automatically match resolving events with an existing alarm.
Alarms with matching resolutions with problems (Ups with Downs), should be indicated with the alarm-type attribute.

* alarm-type="1" (problem alarm)
* alarm-type="2" (resolving alarm)
* alarm-type="3" (notification alarm... alarm with no resolution such as SNMP Authentication Failures)

> == Instantiate new Alarms for existing cleared problem

Reviewer: ```suggestion
== Instantiate new alarms for existing cleared problem
```

----

> Now, with this property set, when a repeat incident occurs and the current state of the Alarm tracking the problem is "Cleared", instead of restating the current Alarm to it's default severity and incrementing the counter, a new instance of the Alarm will be created.

Reviewer: ```suggestion
Now, with this property set, when a repeat incident occurs and the current state of the alarm tracking the problem is "Cleared", instead of restating the current alarm to its default severity and incrementing the counter, a new instance of the alarm will be created.
```

> .New node down Alarm with existing cleared Alarm

Reviewer: ```suggestion
.New node-down alarm with existing cleared alarm
```

.Altered reductionKey
image::alarms/new_after_clear_4.png[]

> == Re-enable legacy dual Alarm state behavior

Reviewer: ```suggestion
== Re-enable legacy dual alarm state behavior
```

#org.opennms.alarmd.legacyAlarmState = false
org.opennms.alarmd.legacyAlarmState = true
----

> NOTE: Setting `legacyAlarmState` will nullify `newIfClearedAlarmExists`

Reviewer: ```suggestion
NOTE: Setting `legacyAlarmState` nullifies `newIfClearedAlarmExists`.
```
